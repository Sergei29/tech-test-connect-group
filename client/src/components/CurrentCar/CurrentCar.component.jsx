import React from "react";
import PropTypes from "prop-types";
import { useCarById } from "../../hooks/useCarById";
//components:
import HeroImage from "./components/HeroImage";
import CarParticulars from "./components/CarParticulars";
import NotFound from "../NotFound";

/**
 * @description functional component for selected car fetched by ID
 * @param {Object} {match} Route prop that passes car ID
 * @returns {JSX} component markup
 */
const CurrentCar = ({ match }) => {
  const selectedId = match.params.id;
  const { loading, carCurrent, error } = useCarById(selectedId);
  const { imgUrl, carName, carSummary, id, meta } = carCurrent;
  const { bodystyles, drivetrain, emissions, passengers } = meta || {
    bodystyles: [],
    drivetrain: [],
    emissions: {},
    passengers: 0,
  };

  if (loading) return <p>Loading...</p>;
  if (error) return <NotFound error={error} />;

  return (
    <div>
      <HeroImage
        imgUrl={imgUrl}
        headingText={`${carName} ${id}`}
        subText={carSummary}
      />
      <CarParticulars
        bodystyles={bodystyles}
        drivetrain={drivetrain}
        emissions={emissions}
        passengers={passengers}
      />
    </div>
  );
};

CurrentCar.propTypes = {
  match: PropTypes.object.isRequired,
};

export default CurrentCar;
