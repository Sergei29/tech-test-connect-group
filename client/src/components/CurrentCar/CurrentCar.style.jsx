import styled from "styled-components";

export const NavContainer = styled.nav`
  display: grid;
  grid-template-columns: auto;
`;
