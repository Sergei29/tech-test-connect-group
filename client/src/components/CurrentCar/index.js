export { default } from "./CurrentCar.component";
