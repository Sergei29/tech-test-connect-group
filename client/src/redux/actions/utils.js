/**
 * @description generates the error message out of received error from api
 * @param {Object} error error received from api
 * @param {String} entity name of the entity: 'carList' or 'carCurrent'
 * @returns {String} aggregated error message
 */
export const getErrorMessage = (error, entity) => {
  let errorMessage = "API call failed.";

  if (error.response && error.response.data) {
    errorMessage = error.response.data;
  } else if (error.message) {
    errorMessage = error.message;
  }

  return `${errorMessage}. ${entity}`;
};
