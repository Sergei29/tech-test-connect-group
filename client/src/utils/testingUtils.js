import React from "react";
import { shallow, mount } from "enzyme";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

const mockStore = configureMockStore([thunk]);

const store = mockStore({
  carList: [
    {
      id: "xe",
      modelYear: "k17",
      url: "/vehicle/xe",
      description:
        "The most advanced, efficient and refined sports saloon that Jaguar has ever produced",
      price: "£30,000",
      media: [{ media: "vehicle", url: "assets/images/xe_k17.jpg" }],
    },
    {
      id: "xf",
      modelYear: "k17",
      url: "/vehicle/xf",
      description:
        "Luxury business saloon with distinctive design, dynamic drive and state-of-the-art technologies.",
      price: "£36,000",
      media: [{ media: "vehicle", url: "assets/images/xf_k17.jpg" }],
    },
  ],
  carCurrent: {
    id: "xe",
    meta: {
      passengers: 5,
      drivetrain: ["AWD", "RWD"],
      bodystyles: ["saloon"],
      emissions: { template: "CO2 Emissions $value g/km", value: 99 },
    },
  },
  error: null,
  loading: false,
});

export const shallowRendererReduxConnected = (ComponentName, props = {}) =>
  shallow(
    <Provider store={store}>
      <BrowserRouter>
        <ComponentName {...props} />
      </BrowserRouter>
    </Provider>
  );

export const mountRendererReduxConnected = (ComponentName, props = {}) =>
  mount(
    <Provider store={store}>
      <BrowserRouter>
        <ComponentName {...props} />
      </BrowserRouter>
    </Provider>
  );

export const shallowRenderer = (ComponentName, props = {}) =>
  shallow(<ComponentName {...props} />);

export const mountRenderer = (ComponentName, props = {}) =>
  mount(
    <BrowserRouter>
      <ComponentName {...props} />
    </BrowserRouter>
  );
