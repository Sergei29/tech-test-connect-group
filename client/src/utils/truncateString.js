export const truncateString = (stringValue, maxLength) => {
  let truncateLocation = stringValue.indexOf(" ", maxLength);
  if (truncateLocation === -1) return stringValue;
  return stringValue.substring(0, truncateLocation) + "...";
};
