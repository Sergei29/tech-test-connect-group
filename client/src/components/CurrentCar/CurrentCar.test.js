import { mountRendererReduxConnected } from "../../utils/testingUtils";
import CurrentCar from "./CurrentCar.component";

describe("CurrentCar", () => {
  let mountWrapper, props;

  beforeEach(() => {
    props = {
      match: { params: { id: "xe" } },
    };
    mountWrapper = mountRendererReduxConnected(CurrentCar, props);
  });

  it("should match the snapshot", () => {
    expect(mountWrapper).toMatchSnapshot();
  });

  it("should contain HeroImage component)", () => {
    expect(mountWrapper.find("HeroImage").length > 0).toBe(true);
  });

  it("should contain CarParticulars component)", () => {
    expect(mountWrapper.find("CarParticulars").length > 0).toBe(true);
  });
});
