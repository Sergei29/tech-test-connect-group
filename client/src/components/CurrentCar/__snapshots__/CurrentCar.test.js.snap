// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrentCar should match the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <CurrentCar
        match={
          Object {
            "params": Object {
              "id": "xe",
            },
          }
        }
      >
        <div>
          <HeroImage
            headingText="Jaguar k17 xe"
            imgUrl="/assets/images/xe_k17.jpg"
            subText="The most advanced, efficient and refined sports saloon that Jaguar has ever produced"
          >
            <styled.div
              imgUrl="/assets/images/xe_k17.jpg"
            >
              <div
                className="sc-bdfBwQ iVhXjw"
              >
                <div
                  className="heroText"
                >
                  <h1>
                    Jaguar k17 xe
                  </h1>
                  <p
                    className="subText"
                  >
                    The most advanced, efficient and refined sports saloon that Jaguar has ever produced
                  </p>
                </div>
              </div>
            </styled.div>
          </HeroImage>
          <CarParticulars
            bodystyles={
              Array [
                "saloon",
              ]
            }
            drivetrain={
              Array [
                "AWD",
                "RWD",
              ]
            }
            emissions={
              Object {
                "template": "CO2 Emissions $value g/km",
                "value": 99,
              }
            }
            passengers={5}
          >
            <styled.div>
              <div
                className="sc-gsTCUz bWpwzB"
              >
                <div
                  className="particularsItem"
                >
                  <span>
                    Bodystyle: 
                  </span>
                  <span>
                    saloon
                  </span>
                </div>
                <div
                  className="particularsItem"
                >
                  <span>
                    Drivetrain: 
                  </span>
                  <span>
                    AWD, RWD
                  </span>
                </div>
                <div
                  className="particularsItem"
                >
                  <span>
                    CO2 Emissions $value g/km
                     :
                  </span>
                  <span>
                    99
                  </span>
                </div>
                <div
                  className="particularsItem"
                >
                  <span>
                    Passengers: 
                  </span>
                  <span>
                    5
                  </span>
                </div>
              </div>
            </styled.div>
          </CarParticulars>
        </div>
      </CurrentCar>
    </Router>
  </BrowserRouter>
</Provider>
`;
